### PR TITLE
iox-#204 making mac semaphore shm compatible

### LINKED ISSUE #204: Unable to run on ROS2 demo application on OSX
## Required information

**Operating system:**
OSX

**Compiler version:**
Apple LLVM version 10.0.1 (clang-1001.0.46.4)

**Observed result or behavior:**
segmentation fault

```
1970-01-13 04:30:05.895 [ Debug ]: Application registered management segment 0x103000000 with size 71578784 to id 1
1970-01-13 04:30:05.897 [ Info  ]: Application registered payload segment 0x110000000 with size 149655680 to id 2
Process 74700 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7ffc1be007c0)
    frame #0: 0x000000010209af90 librmw_iceoryx_name_conversion.dylib`std::__1::__atomic_base<int, false>::store(this=0x00007ffc1be007c0, __d=0, __m=memory_order_relaxed) at atomic:921:10
   918 	    _LIBCPP_INLINE_VISIBILITY
   919 	    void store(_Tp __d, memory_order __m = memory_order_seq_cst) _NOEXCEPT
   920 	      _LIBCPP_CHECK_STORE_MEMORY_ORDER(__m)
-> 921 	        {__c11_atomic_store(&__a_, __d, __m);}
   922 	    _LIBCPP_INLINE_VISIBILITY
   923 	    _Tp load(memory_order __m = memory_order_seq_cst) const volatile _NOEXCEPT
   924 	      _LIBCPP_CHECK_LOAD_MEMORY_ORDER(__m)
Target 0: (talker) stopped.
```

bt:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7ffc1be007c0)
  * frame #0: 0x000000010209af90 librmw_iceoryx_name_conversion.dylib`std::__1::__atomic_base<int, false>::store(this=0x00007ffc1be007c0, __d=0, __m=memory_order_relaxed) at atomic:921:10
    frame #1: 0x000000010209aef0 librmw_iceoryx_name_conversion.dylib`iox_sem_init(sem=0x0000000106365710, (null)=1, value=0) at semaphore.cpp:222:19
    frame #2: 0x00000001020683b2 librmw_iceoryx_name_conversion.dylib`iox::cxx::SmartC<int (iox_sem_t*, int, unsigned int), int, iox_sem_t*, int, unsigned int>::SmartC(this=0x00007ffeefbece50, file="/Users/karsten/workspace/ros2/iceoryx_ws/osx/src/iceoryx/iceoryx_utils/source/posix_wrapper/semaphore.cpp", line=243, func="bool iox::posix::Semaphore::init(iox_sem_t *, const int, const unsigned int)", f_function=0x000000010209aeb0, f_mode=0x00007ffeefbece34, f_returnValues=0x00007ffeefbece20, f_ignoredValues=0x00000001020a8f30, f_args=0x0000000106365710, f_args=1, f_args=0)(iox_sem_t*, int, unsigned int), iox::cxx::ReturnMode const&, std::initializer_list<int> const&, std::initializer_list<int> const&, iox_sem_t*, int, unsigned int) at smart_c.inl:99:21
    frame #3: 0x000000010206823d librmw_iceoryx_name_conversion.dylib`iox::cxx::SmartC<int (iox_sem_t*, int, unsigned int), int, iox_sem_t*, int, unsigned int>::SmartC(this=0x00007ffeefbece50, file="/Users/karsten/workspace/ros2/iceoryx_ws/osx/src/iceoryx/iceoryx_utils/source/posix_wrapper/semaphore.cpp", line=243, func="bool iox::posix::Semaphore::init(iox_sem_t *, const int, const unsigned int)", f_function=0x000000010209aeb0, f_mode=0x00007ffeefbece34, f_returnValues=0x00007ffeefbece20, f_ignoredValues=0x00000001020a8f30, f_args=0x0000000106365710, f_args=1, f_args=0)(iox_sem_t*, int, unsigned int), iox::cxx::ReturnMode const&, std::initializer_list<int> const&, std::initializer_list<int> const&, iox_sem_t*, int, unsigned int) at smart_c.inl:101:1
    frame #4: 0x000000010206414c librmw_iceoryx_name_conversion.dylib`iox::cxx::SmartC<int (iox_sem_t*, int, unsigned int), int, iox_sem_t*, int, unsigned int> iox::cxx::makeSmartCImpl<int (file="/Users/karsten/workspace/ros2/iceoryx_ws/osx/src/iceoryx/iceoryx_utils/source/posix_wrapper/semaphore.cpp", line=243, func="bool iox::posix::Semaphore::init(iox_sem_t *, const int, const unsigned int)", f_function=0x000000010209aeb0, f_mode=0x00007ffeefbece34, f_returnValues=0x00007ffeefbece20, f_ignoredValues=0x00000001020a8f30, f_args=0x0000000106365710, f_args=1, f_args=0), int, iox_sem_t*, int, unsigned int>(char const*, int, char const*, int  const(&)(iox_sem_t*, int, unsigned int), iox::cxx::ReturnMode const&, std::initializer_list<int> const&, std::initializer_list<int> const&, iox_sem_t*, int, unsigned int) at smart_c.inl:32:19
    frame #5: 0x00000001020632ed librmw_iceoryx_name_conversion.dylib`iox::posix::Semaphore::init(this=0x00007ffeefbecfe0, handle=0x0000000106365710, pshared=1, value=0) at semaphore.cpp:243:13
    frame #6: 0x0000000102063430 librmw_iceoryx_name_conversion.dylib`iox::posix::Semaphore::Semaphore(this=0x00007ffeefbecfe0, handle=0x0000000106365710, value=0) at semaphore.cpp:200:9
    frame #7: 0x0000000102063493 librmw_iceoryx_name_conversion.dylib`iox::posix::Semaphore::Semaphore(this=0x00007ffeefbecfe0, handle=0x0000000106365710, value=0) at semaphore.cpp:199:1
    frame #8: 0x000000010202a662 librmw_iceoryx_name_conversion.dylib`iox::cxx::expected<iox::posix::Semaphore, iox::posix::SemaphoreError> DesignPattern::Creation<iox::posix::Semaphore, iox::posix::SemaphoreError>::create<iox_sem_t*, int>(args=0x00007ffeefbed0d8, args=0x00007ffeefbed0d4) at creation.hpp:31:23
    frame #9: 0x000000010202a4fe librmw_iceoryx_name_conversion.dylib`iox::popo::ReceiverPort::GetShmSemaphore(this=0x0000000100b00fd8) at receiver_port.cpp:290:40
    frame #10: 0x0000000100d2249c librmw_iceoryx_cpp.dylib`iox::popo::Subscriber_t<iox::popo::ReceiverPort>::getSemaphore(this=0x0000000100b00e60) const at subscriber.inl:222:33
    frame #11: 0x0000000100d22046 librmw_iceoryx_cpp.dylib`::rmw_create_wait_set(context=0x0000000100c01420, max_conditions=2) at rmw_wait_set.cpp:66:33
    frame #12: 0x000000010033200f librcl.dylib`rcl_wait_set_init + 511
    frame #13: 0x0000000100146d2b librclcpp.dylib`rclcpp::Executor::Executor(rclcpp::ExecutorOptions const&) + 779
    frame #14: 0x000000010014e66e librclcpp.dylib`rclcpp::executors::SingleThreadedExecutor::SingleThreadedExecutor(rclcpp::ExecutorOptions const&) + 14
    frame #15: 0x000000010000eb52 talker`main + 322
    frame #16: 0x00007fff5d77f3d5 libdyld.dylib`start + 1
    frame #17: 0x00007fff5d77f3d5 libdyld.dylib`start + 1
```

**Expected result or behaviour:**
`RMW_IMPLEMENTATION=rmw_iceoryx_cpp ros2 run demo_nodes_cpp talker` to run successfully.

### LINKED ISSUE #407: macos build with `with-dds-gateway` fails (github workflow )
## Required information

**Operating system:**
macos-10.15

**Compiler version:**
AppleClang 12.0.0.12000032

**Observed result or behaviour:**
macos build fails in github workflow job (when `with-dds-gateway` is enabled)
```
[ 33%] Building CXX object CMakeFiles/iceoryx_dds_messages.dir/Mempool.cpp.o
ANTLR Tool version 4.4 used for code generation does not match the current runtime version 4.5ANTLR Tool version 4.4 used for code generation does not match the current runtime version 4.5In file included from /Users/runner/work/iceoryx/iceoryx/build/iceoryx_dds_messages/Mempool.cpp:1:
In file included from /Users/runner/work/iceoryx/iceoryx/build/iceoryx_dds_messages/Mempool.h:4:
In file included from /Users/runner/work/iceoryx/iceoryx/build/dependencies/install/include/ddscxx/dds/core/ddscore.hpp:22:
In file included from /Users/runner/work/iceoryx/iceoryx/build/dependencies/install/include/ddscxx/dds/core/types.hpp:28:
In file included from /Users/runner/work/iceoryx/iceoryx/build/dependencies/install/include/ddscxx/dds/core/macros.hpp:22:
/Users/runner/work/iceoryx/iceoryx/build/dependencies/install/include/ddscxx/dds/core/detail/macros.hpp:88:14: fatal error: 'boost/static_assert.hpp' file not found
#    include <boost/static_assert.hpp>
             ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/iceoryx_dds_messages.dir/Mempool.cpp.o] Error 1
make[1]: *** [CMakeFiles/iceoryx_dds_messages.dir/all] Error 2
make: *** [all] Error 2
```

**Expected result or behaviour:**
Successful compilation & test run

**Conditions where it occurred / Performed steps:**
Enable option `with-dds-gateway` while invoking macos  build job in github workflow(using `iceoryx_build_test.sh`)

### PR DESCRIPTION
**Pull request by elfenpiff**
Signed-off-by: Christian Eltzschig <me@elchris.org>

## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer
The shm incompatiblity was **partly** caused by the member `m_value` in the mac version of `iox_sem_t`. Removing it and adding some comments on why heap usage is still allowed in some places should fix the issue.
**Update**: The main problem is that dispatch_semaphore_create() in iox_sem_init() returns an address in RouDi's address space that is not accessible from an app (used by dispatch_semaphore_wait()). This is **not** resolved in this PR.

**Update**: Since the dispatch_semaphore_create() was not shared memory compatible we switched to a combination of an inter process pthread_mutex and pthread_cond to simulate the semaphore behaviour. All examples are running again without seg faults. And as a Christmas goodie goodie: DDS is now also working on Mac OS!

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [ ] All checkboxes in the PR checklist are checked or crossed out
1. [ ] Merge

## References

- closes #204 
- closes #407

### CODE REVIEW

**Review by mossmaurice — DISMISSED** [review]
LGTM, maybe you can disable the failing test in MacOS santize build

Have you verified the ROS2 example from #204?

**Comment by marthtz** [comment]
Although moving `m_value` from heap to stack might solve an issue it doesn't resolve the shared memory semaphore issue unfortunately.
The problem is that `dispatch_semaphore_create()` in `iox_sem_init()` returns an address in RouDi's address space that is not accessible from an app (used by `dispatch_semaphore_wait()`).
So, macOS support is still broken and #204 not resolved.

**Review by marthtz — DISMISSED** [review]
#204 cannot be closed after this PR is merged as the main problem is still not resolved.

**Comment by elfenpiff** [comment]
@marthtz could you please rerun the ROS2 example. I think everything should work now!

File: iceoryx_dds/CMakeLists.txt [hunk 1]
```diff
@@ -133,7 +142,10 @@ target_link_libraries(iox-gw-dds2iceoryx
     iceoryx_posh::iceoryx_posh
     ${PROJECT_NAMESPACE}::iceoryx_dds
 )
-
+set_target_properties(iox-gw-dds2iceoryx PROPERTIES
```
Comment (dkroenke) [inline iceoryx_dds/CMakeLists.txt:145]: You can safe code by adding multiple targets to it and delete the two statements above
```
set_target_properties(iceoryx_dds iox-gw-iceoryx2dds iox-gw-dds2iceoryx PROPERTIES
    CXX_STANDARD_REQUIRED ON
    CXX_STANDARD ${ICEORYX_CXX_STANDARD}
)
```

File: iceoryx_utils/platform/mac/include/iceoryx_utils/platform/semaphore.hpp [hunk 2]
```diff
@@ -31,11 +31,16 @@ struct iox_sem_t
     union
     {
         sem_t* posix;
-        dispatch_semaphore_t dispatch;
+
+        struct
+        {
+            pthread_mutex_t mtx;
+            pthread_cond_t variable;
+        } condition;
```
Comment (elBoberido) [inline iceoryx_utils/platform/mac/include/iceoryx_utils/platform/semaphore.hpp:39]: with this, the semaphore definitely must not be movable

**Review by elBoberido — APPROVED** [review]

**Comment by elBoberido** [comment]
I think it's undefined if the mutex is move, but at least mac works now. We shouldn't move the semaphore though.

File: iceoryx_dds/CMakeLists.txt (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_dds/CMakeLists.txt]: done

**Comment by elfenpiff** [comment]
@marthtz I just fixed the iceperf example in here.

**Review by dkroenke — APPROVED** [review]

**Review by mossmaurice — APPROVED** [review]

**Pull request merged**